File: src/main/java/dev/External/Tools/RequirementsTracker.java
package dev.External.Tools;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class RequirementsTracker
{
    public static void runScan() throws IOException
    {
        Path sourceDir = Paths.get("src");
        Path requirementsFile = Paths.get("src/main/resources/requirements.txt");

        List<String> requirements = loadRequirements(requirementsFile);
        Set<String> foundRequirements = new HashSet<>();


        Files.walk(sourceDir)
                .filter(path -> path.toString().endsWith(".java"))
                .forEach(path ->
                {
                    try
                    {
                        processFile(path, requirements, foundRequirements);
                    } catch (IOException e)
                    {
                        System.err.println("Error on reading from file: " + path);
                    }
                });

        printResults(requirements, foundRequirements);
    }

    private static List<String> loadRequirements(Path file) throws IOException
    {
        return Files.lines(file)
                .map(String::trim)
                .filter(line -> !line.isEmpty() && !line.startsWith("#"))
                .collect(Collectors.toList());
    }

    private static void processFile(Path filePath, List<String> requirements, Set<String> foundRequirements) throws IOException
    {
        List<String> lines = Files.readAllLines(filePath);
        String REQ_COMMENT_PREFIX = "// Req. Nr.: "; // --ignore

        for (int lineNumber = 0; lineNumber < lines.size(); lineNumber++)
        {
            String line = lines.get(lineNumber);
            if (line.contains(REQ_COMMENT_PREFIX))
            {
                if (line.contains("--ignore"))
                {
                    continue;
                }
                String reqNumber = line.substring(line.indexOf(REQ_COMMENT_PREFIX) + REQ_COMMENT_PREFIX.length()).trim();
                String reqEntry = getRequirementText(reqNumber, requirements);
                String combinedEntry = String.format("%s | File: %s | Line: %d", reqEntry, filePath, lineNumber + 1);

                foundRequirements.add(combinedEntry);
            }
        }
    }

    private static String getRequirementText(String reqNumber, List<String> requirements)
    {
        return requirements.stream()
                .filter(req -> req.startsWith(reqNumber + "."))
                .findFirst()
                .orElse("Unknown Requirement");
    }

    private static void printResults(List<String> requirements, Set<String> foundRequirements)
    {
        System.out.println("Found Requirements:");
        foundRequirements.forEach(System.out::println);

        System.out.println("\nNot found Requirements:");
        requirements.stream()
                .filter(req -> foundRequirements.stream().noneMatch(found -> found.contains(req)))
                .forEach(System.out::println);
    }
}

File: Backend/tools/RequirementsTracker.py
import os
from pathlib import Path


def run_scan():
    source_dir = Path("../src")
    requirements_file = Path("../src/main/resources/requirements.txt")

    requirements = load_requirements(requirements_file)
    found_requirements = set()

    for root, _, files in os.walk(source_dir):
        for file in files:
            if file.endswith(".java"):
                path = Path(root) / file
                try:
                    process_file(path, requirements, found_requirements)
                except IOError:
                    print(f"Error on reading from file: {path}")

    print_results(requirements, found_requirements)


def load_requirements(file):
    with open(file, encoding="utf-8") as f:
        return [line.strip() for line in f if line.strip() and not line.startswith("#")]


def process_file(file_path, requirements, found_requirements):
    req_comment_prefix = "// Req. Nr.: "  # --ignore

    with open(file_path, encoding="utf-8") as f:
        lines = f.readlines()

    for line_number, line in enumerate(lines, start=1):
        if req_comment_prefix in line:
            if "--ignore" in line:
                continue
            req_number = line.split(req_comment_prefix, 1)[1].strip()
            req_entry = get_requirement_text(req_number, requirements)
            combined_entry = f"{req_entry} | File: {file_path} | Line: {line_number}"
            found_requirements.add(combined_entry)


def get_requirement_text(req_number, requirements):
    for req in requirements:
        if req.startswith(f"{req_number}."):
            return req
    return "Unknown Requirement"


def print_results(requirements, found_requirements):
    print("Found Requirements:")
    for req in found_requirements:
        print(req)

    print("\nNot found Requirements:")
    for req in requirements:
        if not any(req in found for found in found_requirements):
            print(req)


if __name__ == "__main__":
    run_scan()
